File: 2025/python/day11.py
#!/usr/bin/env python3

from dataclasses import dataclass
from aoc import *

@dataclass
class Node:
    name: str
    inputs: list[str]
    outputs: list[str]
    paths: int
    depth: int

def parse_input(input: str) -> dict[str, Node]:
    nodes = {}
    for line in input.splitlines():
        parts = line.split()
        name = parts[0][:-1]
        node = nodes.get(name)
        if not node:
            node = Node(name, [], [], 0, 0)
            nodes[name] = node
        node.outputs = parts[1:]
        for output in node.outputs:
            out = nodes.get(output)
            if not out:
                out = Node(output, [], [], 0, 0)
                nodes[output] = out
            out.inputs.append(node.name)
    return nodes

def find_paths(start: str, end: str, nodes: dict[str, Node]) -> int:
    for n in nodes.values():
        n.paths = 0
    in_progress = set(nodes[start].outputs)
    queue = list(in_progress)
    nodes[start].paths = 1
    while queue:
        name = queue.pop(0)
        in_progress.remove(name)
        node = nodes[name]
        paths = sum([nodes[inp].paths for inp in node.inputs])
        if paths != node.paths:
            node.paths = paths
            for inp in node.outputs:
                if inp not in in_progress:
                    in_progress.add(inp)
                    queue.append(inp)
    return nodes[end].paths

def part1(input: str):
    return 0
    nodes = parse_input(input)
    return find_paths('you', 'out', nodes)


def part2(input: str):
    nodes = parse_input(input)
    dac = find_paths('svr', 'dac', nodes)
    fft = nodes['fft'].paths
    print(f"Paths to DAC: {dac}, Paths to FFT: {fft}")
    if dac < fft:
        middle = find_paths('dac', 'fft', nodes)
        print(f"Paths from DAC to FFT: {middle}")
        return dac * middle * find_paths('fft', 'out', nodes)
    else:
        middle = find_paths('fft', 'dac', nodes)
        print(f"Paths from FFT to DAC: {middle}")
        print(find_paths('dac', 'out', nodes))
        return dac * middle * find_paths('dac', 'out', nodes)

# 207000528761440412149500
if __name__ == '__main__':
    main()

File: 2025/go/day11.go
package main

import (
	"github.com/bpa/aoc"
	"github.com/gammazero/deque"
)

type Day11Node struct {
	name     string
	input    []int
	output   []int
	paths    int
	enqueued bool
}

func newDay11Node(name string) *Day11Node {
	return &Day11Node{
		name:   name,
		input:  make([]int, 0, 16),
		output: make([]int, 0, 16),
	}
}

func newDay11Graph(input string) aoc.IndexedArray[string, *Day11Node] {
	graph := aoc.NewIndexedArrayFunc(1024, newDay11Node)
	for _, line := range aoc.Delimited(input) {
		name := line[0]
		name = name[:len(name)-1]
		i, input := graph.GetOrCreate(name)
		for _, c := range line[1:] {
			o, output := graph.GetOrCreate(c)
			input.output = append(input.output, o)
			output.input = append(output.input, i)
		}
	}
	return graph
}

func day11Dist(graph *aoc.IndexedArray[string, *Day11Node], from, to string, clear bool) int {
	f, ok := graph.Ind[from]
	if !ok {
		return 0
	}

	if clear {
		for _, n := range graph.Data {
			n.paths = 0
		}
	}
	graph.Data[f].paths = 1

	var queue deque.Deque[int]
	queue.SetBaseCap(1024)
	queue.CopyInSlice(graph.Data[f].output)
	for queue.Len() > 0 {
		i := queue.PopFront()
		node := graph.Data[i]
		total := 0
		for _, j := range node.input {
			total += graph.Data[j].paths
		}
		node.paths = total
		for _, j := range node.output {
			out := graph.Data[j]
			if !(out.enqueued) {
				out.enqueued = true
				queue.PushBack(j)
			}
		}
		node.enqueued = false
	}
	_, t := graph.GetOrCreate(to)
	return t.paths
}

func day11Part1(input string) int {
	graph := newDay11Graph(input)
	return day11Dist(&graph, "you", "out", false)
}

func day11Part2(input string) int {
	graph := newDay11Graph(input)
	dac := day11Dist(&graph, "svr", "dac", false)
	_, fftNode := graph.GetOrCreate("fft")
	fft := fftNode.paths
	if dac < fft {
		middle := day11Dist(&graph, "dac", "fft", true)
		return dac * middle * day11Dist(&graph, "fft", "out", true)
	} else {
		middle := day11Dist(&graph, "fft", "dac", true)
		return fft * middle * day11Dist(&graph, "dac", "out", true)
	}
}

func main() {
	aoc.Main(day11Part1, day11Part2)
}
